### ISSUE TITLE
Main API tries and fails to load a dependency that's missing pom.xml

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon.language#625]
See https://github.com/ceylon/ceylon-module-resolver/issues/113 for details. (The precipitating factor in that issue of erroneously including optional dependencies has been fixed, but not the inability to load jsr305 in general.)

To reproduce, attempt to use the main api or the tool provider api to load a module that includes:

```ceylon
module simple "1.0.0" {
  import "com.google.code.findbugs:jsr305" "1.3.9";
}
```

**Comment by jvasileff** [comment]
So it looks like about 40% of the jars in my `~/.m2` directory do not have `pom.xml`, which I guess also means that 40% of them were created by sane people who don't build with mvn. Here's a few:

```ceylon
module simple "1.0.0" {
  import "log4j:log4j" "1.2.14";
  import "junit:junit" "4.12";
  import "postgresql:postgresql" "9.1-901-1.jdbc4";
  import "org.springframework:spring-core" "4.0.3.RELEASE";
  import "org.apache.ant:ant" "1.8.2";
  import "javax.mail:mail" "1.4.1";
  import "xalan:xalan" "2.7.1";
}
```

**Comment by quintesse** [comment]
To be honest I don't know what would be the best way forward with this, it seems obvious that once you have a flat classpath you can't expect the meta data to always be available. So I can think of two things:

 - require you to specify that data the same way we do now: have `module.xml` or `module.properties` files that specify the required information (maybe we could extend that to `pom.xml` files as well).
 - and have an option to say: please, just trust me that all dependencies will be on the classpath. This of course means having no knowledge about those dependencies in the meta model.

**Comment by gavinking** [comment]
I think it's clear that those "flat classpath" jars should all be deployed on a single shared classloader, and that we need a way to declare that. Perhaps it's even something that happens by default when the needed metadata is missing.

**Comment by quintesse** [comment]
@jvasileff I added the possibility of ignoring missing modules to the API (`--allow-missing-modules` on the command line, `Main.instance().allowMissingModules(true).run(....)` in the API itself). Could you test if that solves this issue for you?

**Comment by jvasileff** [comment]
Cool, thanks, I'll check it out. Although it will probably not be for a couple days.

I'm assuming that this will help for completely missing modules that we don't want, but for "provided" modules or substitutions, there may be runtime issues as in https://github.com/ceylon/ceylon.language/pull/626?

This is really just a question, a bit beyond the scope of this particular issue. But it occurs to me that with the war tool, the runtime may not be aware of all of the modules that are actually being used, due to provided, or even overrides.xml, which I don't think is available right now at runtime, in wars.

**Comment by jvasileff** [comment]
Actually, that was probably a poorly formed question :). If you can infer what I meant, great! Otherwise I'll try to ask it again in a better way, later.

**Comment by quintesse** [comment]
This *only* provides for an option to ignore dependencies that Ceylon's meta model thinks should be there but can't find (for example because the needed information is missing from the jar's `META-INF`). In fact it will now even ignore dependencies that are *actually* missing from the classpath!
It's just an option that says "trust me, I know what I'm doing".

**Comment by jvasileff** [comment]
I tried this out, and it does indeed suppress the error. But I'm not sure that is a good thing, since instead, the module goes unloaded (as mentioned.) For clarity, a quick demo of the impact:

```ceylon
import org.apache.log4j {
    Level
}
import ceylon.language.meta {
    type
}

shared void run() {
    Anything all = Level.\iALL; // ok, proves Level was loaded by the JVM
    print(all is Level); // ditto

    Anything levels = {Level.\iALL};
    print(levels is {Level*}); // error, ModelResolutionException
}
```

I guess this starts to butt up against some the other ideas of *fully* supporting a trust me or ad-hoc mode, but it seems like it *might* be possible to discuss this issue independently of that, since this affects the more narrow case where 1) all dependencies are known and resolved cleanly at compile time based on the module descriptor and overrides file, and 2) we are willing to provide a properly organized flat repo at runtime.

But I do understand, a problem is that there isn't enough information to rebuild the dependency tree at runtime when poms are missing. Even if it were possible to "force load" log4j, we wouldn't know what dependencies of log4j to load.

Maybe there should be something like `ceylon classpath` that lists all of the jar file names to force-load as modules. IDK.

Trying not to lose site of the end goal of running Ceylon apps in a various environments, a few additional items come to mind:

- Support for jars w/o poms in *both* the main api and tool provider api
- Support for overrides.xml at runtime in the main api and tool provider api (assuming 1. runtime metatdata is not somehow made orthogonal to dependency resolution, and 2. overrides.xml would be used at all, instead of the pre-determined list of jar/module names idea)
- Support for ^ in the war initializer

**Comment by quintesse** [comment]
> that lists all of the jar file names to force-load as modules

Well that would assume that there is always a one to one mapping of jars and modules, or that there are even jars and not just a bunch of class files in a directory.
To me that just seems like generating metadata for the heck of it.

Wouldn't it be a much simpler solution to just put everything that we can't match to a module into a catch-all kitchen-sink module?

That doesn't mean I'm against trying to provide proper metadata, but it seems like an orthogonal problem. Especially because there will be no way to be sure, it will just be another, more elaborate case of "please trust me".

**Comment by jvasileff** [comment]
Definitely. That sounds like a different, but more general solution. It might also help with classes that "come out of nowhere", as in the example above, but if `log4j` were loaded at the container level, out of reach of the current `jar` scanner.

**Comment by jvasileff** [comment]
One thing that I think is true with any of these scenarios, is that without knowledge of the dependency tree, these classes will have to be assumed to both import and be imported by all other modules\*. (Technically, I suppose only modules with unresolved imports would have to import the catch-all bucket.) Roughly, I think it's a case, as in https://github.com/ceylon/ceylon.language/pull/626, where the visibility metadata used at runtime for type checking can be in conflict with actual loaded classes inside the JVM. 

(\*) Conceptually; the PR handles this by not allowing import metadata to affect subtyping, which seems to work well and afaict be the "right thing to do" at runtime, but may disable some valuable compile time checks.

**Comment by FroMage** [comment]
Moving to 1.3.
